[Pauli4] PauliWord and PauliSentence Documentation (#5017)
PennyLaneAI/pennylane: doc/code/qml_pauli.rst

instantiated by passing a dictionary of wires and their associated Pauli operators.

>>> from pennylane.pauli import PauliWord
>>> pw1 = PauliWord({0:"X", 1:"Z"})  # X@Z
>>> pw2 = PauliWord({0:"Y", 1:"Z"})  # Y@Z
>>> pw1, pw2
(X(0) @ Z(1), Y(0) @ Z(1))

The purpose of this class is to efficiently compute products of Pauli words and
obtain the matrix representation.

> >>> pw1 @ pw2

Author: Yeah I agree with all that, I was also a bit confused. I wasnt part of the development of the original class, so can only guess what the reasons were:
- returning the tuple `(op_res, coeff)` is slightly more lightweight than returning a `PauliSentence`
```
def pauli_matmul(pw1, pw2):
    op_res, coeff = pw1 @ pw2
    return PauliSentence({op_res: coeff})
>>> %timeit pw1 @ pw2 
740 ns ± 16.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
>>> %timeit pauli_matmul(pw1, pw2)
973 ns ± 19.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
- For some arithmetic you may want to stay with just "pure" pauliwords and tracking coefficients separately is not an issue. This may be because of performance considerations (see point above), or that linear independence between PauliWords is just equality (up to a factor). Think PauliWords are the canonical basis `(1,0,0..)`, `(0, 1, 0..)`, `(0, 0, 1,..)` vs PauliSentence are arbitrary vectors `(a, b, c, ..)`, ... and you sometimes just wanna stay in the canonical basis. That being said I agree that from a user perspective its odd to not get a `PauliSentence`
- I think a lot of focus / motivation for this class in the first place was efficiently creating (sparse) matrices, hence the focus on that in the docs.

I think these are all valid concerns, maybe we move them to a separate PR/forum though?

> >>> pw1 @ pw2

Reviewer: Exactly. The motivation here was "fast", which meant lightweight.  And I honestly didn't even think about it.

I'd be fine with switching to having `@` return a pauli sentence, as long as we have a different method for the more lightweight version that can be used in big problems.

> >>> pw1 @ pw2

Author: Lets move this discussion to this PR https://github.com/PennyLaneAI/pennylane/pull/5054
But essentially this is doing as you propose: Have `@` return a paulisentence but offer a different method for more lightweight processing :-)

(Z(0), 1j)
>>> pw1.to_mat(wire_order=[0, 1])
array([[ 0,  0,  1,  0],
       [ 0,  0,  0, -1],
       [ 1,  0,  0,  0],
       [ 0, -1,  0,  0]])

The :class:`~pennylane.pauli.PauliSentence` class represents linear combinations of
Pauli words. Using a similar dictionary based approach we can efficiently add, multiply
and extract the matrix of operators in this representation.

>>> ps1 = PauliSentence({pw1: 1.2, pw2: 0.5j})

> >>> ps2 = PauliSentence({pw1: -1.2})

Author: I think its actually okay to have this block explore addition and multiplication. The text I would think of to add in between sounds like null-text "We can add two Pauli words:", "We can multiply two Pauli words:", .. which is quite clear from just looking at the code I think. Do you have a concrete nicer idea?

> >>> ps2 = PauliSentence({pw1: -1.2})

Reviewer 2: I agree it doesn't need any text, it's pretty self-explanatory (or already explained by the existing text) in that regard. I just meant a blank line after each output so that it renders as a few cells with white space between, so it's easier to skim.

> >>> ps2 = PauliSentence({pw1: -1.2})

Author: Ah that sounds good 👍 
✅

       [ 0,  0,  0, -1],
       [ 1,  0,  0,  0],
       [ 0, -1,  0,  0]])

The :class:`~pennylane.pauli.PauliSentence` class represents linear combinations of
Pauli words. Using a similar dictionary based approach we can efficiently add, multiply
and extract the matrix of operators in this representation.

>>> ps1 = PauliSentence({pw1: 1.2, pw2: 0.5j})
>>> ps2 = PauliSentence({pw1: -1.2})
>>> ps1
1.2 * X(0) @ Z(1)
+ 0.5j * Y(0) @ Z(1)

>>> ps1 + ps2
0.0 * X(0) @ Z(1)
+ 0.5j * Y(0) @ Z(1)

>>> ps1 @ ps2
-1.44 * I
+ (-0.6+0j) * Z(0)

>>> (ps1 + ps2).to_mat(wire_order=[0, 1])
array([[ 0. +0.j,  0. +0.j,  0.5+0.j,  0. +0.j],
       [ 0. +0.j,  0. +0.j,  0. +0.j, -0.5+0.j],
       [-0.5+0.j,  0. +0.j,  0. +0.j,  0. +0.j],
       [ 0. +0.j,  0.5+0.j,  0. +0.j,  0. +0.j]])

We can intuitively use Pauli arithmetic to construct Hamiltonians consisting of :class:`~pennylane.pauli.PauliWord`
and :class:`~pennylane.pauli.PauliSentence` objects like the spin-1/2 XXZ model Hamiltonian,

.. math:: H_\text{XXZ} = \sum_j [J^\bot (X_j X_{j+1} + Y_j Y_{j+1}) + J^\text{ZZ} Z_j Z_{j+1} + h Z_j].

Here we look at the simple topology of a one-dimensional chain with periodic boundary conditions

> (i.e. qubit number :math:`n \equiv 0` for pythonic numbering of wires, e.g. `[0, 1, 2, 3]` for `n=4`).

Reviewer 2: ```suggestion
(i.e. qubit number :math:`n \equiv 0` for pythonic numbering of wires, e.g. ``[0, 1, 2, 3]`` for ``n=4``).
```

       [-0.5+0.j,  0. +0.j,  0. +0.j,  0. +0.j],
       [ 0. +0.j,  0.5+0.j,  0. +0.j,  0. +0.j]])

We can intuitively use Pauli arithmetic to construct Hamiltonians consisting of :class:`~pennylane.pauli.PauliWord`
and :class:`~pennylane.pauli.PauliSentence` objects like the spin-1/2 XXZ model Hamiltonian,

.. math:: H_\text{XXZ} = \sum_j [J^\bot (X_j X_{j+1} + Y_j Y_{j+1}) + J^\text{ZZ} Z_j Z_{j+1} + h Z_j].

> Here we look at the simple topology of a one-dimensional chain with periodic boundary conditions (i.e. qubit number :math:`n+1 \equiv 1`).

Reviewer 2: Just a preference but since if you print `H` it uses wires [0, 1, 2, 3] instead of [1, 2, 3, 4], I think the boundary condition makes more sense to express as $n \equiv 0$.

Author: Sounds good ✅

In code we can do this via the following example with 4 qubits.

.. code-block:: python3

    n_wires = 4
    J_orthogonal = 1.5

>     ops = [J_orthogonal * (PauliWord({i:"X", (i+1)%n_wires:"X"}) + PauliWord({i:"Y", (i+1)%n_wires:"Y"})) for i in range(n_wires)]

Reviewer 2: This line is much longer than the code block in the docs - could we break it at the `+`, or something similar?

Author: ✅

       [-0.5+0.j,  0. +0.j,  0. +0.j,  0. +0.j],
       [ 0. +0.j,  0.5+0.j,  0. +0.j,  0. +0.j]])

> We can intuitively use pauli arithmetic to construct Hamiltonians consisting of PauliWords and PauliSentences like

Reviewer 3: ```suggestion
We can intuitively use Pauli arithmetic to construct Hamiltonians consisting of ``PauliWord`` and ``PauliSentence`` objects like
```

the spin-1/2 XXZ model Hamiltonian,

.. math:: H_\text{XXZ} = \sum_j [J^\bot (X_j X_{j+1} + Y_j Y_{j+1}) + J^\text{ZZ} Z_j Z_{j+1} + h Z_j].

> Here we look at the simple topology of a on a one-dimensional chain with periodic boundary conditions (i.e. qubit number :math:`n+1 \equiv 1`).

Reviewer 3: ```suggestion
Here we look at the simple topology of a one-dimensional chain with periodic boundary conditions (i.e. qubit number :math:`n+1 \equiv 1`).
```
